### ISSUE TITLE
How to get the pruned attention module

### ISSUE THREAD

**Issue by sdc17** [Arnav0400/ViT-Slim#5]
Hello, thanks for your nice work!
As stated in your paper, each head in the attention module may have different dimension, but I have no idea how to get the pruned attention module. I mean the really compressed module that we don't need to use the 0-1 tensor to mask part of the input.
```python
class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0., num_patches=197):
        super().__init__()
        self.num_heads = num_heads
        self.num_patches = num_patches
        head_dim = dim // num_heads
        self.scale = qk_scale or head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
        q, k, v = qkv[0], qkv[1], qkv[2]   # make torchscript happy (cannot use tensor as tuple)

        attn = (q @ k.transpose(-2, -1)) * self.scale
        attn = attn.softmax(dim=-1)
        attn = self.attn_drop(attn)

        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
        x = self.proj(x)
        x = self.proj_drop(x)
        return x
```
As shown in the `forward` of the `Attention`, the compressed `q` should still be legal tensor, which means every head in the `q` should have the same dimension. And this is contradictory to the results of search that each head may have different dimension.
I have checked the main and throughput branch of this repo, but haven't found the relevant code that actually compress the attention module. I would appreciate it if you could point out where the relevant code is located or provide the relevant code. 
Thanks for your reading and look forward to your reply.

**Comment by yu868** [comment]
Yes, I think so. The reported code does not capture the pruning model structure described in the article. In my opinion, this article only implements logical pruning, or the meaning of the article is not reflected in the code, and does not achieve the so-called effect of reducing the number of parameters. Of course, the idea and writing of the article are quite good, ha ha ha

**Comment by Arnav0400** [comment]
MLP and Patch dimensions bring in a direct improvement in terms of practical efficiency without any specific hardware tuning. For the MHSA module, the major computation is spread between qkv linear layer, projection linear layer, and attention matrix multiplication. The two linear layers can be directly parallelized for all heads together without any tuning, but to extract optimal efficiency, attention matrix multiplication of different heads would need specific tuning. Overall, with or without specific tuning, better practical efficiency can be obtained by our method.
